src/Sulu/Bundle/PageBundle/Tests/Functional/Reference/Provider/PageReferenceProviderTest.php

```diff
@@ -0,0 +1,125 @@
+    }
+
+    public function testUpdateReferences(): void
+    {
+        /** @var SnippetDocument $snippet */
+        $snippet = $this->documentManager->create('snippet');
+        $snippet->setStructureType('animal');
+        $snippet->setTitle('Example snippet');
+        $snippet->setWorkflowStage(WorkflowStage::PUBLISHED);
+        $this->documentManager->persist($snippet, 'en');
+
+        /** @var PageDocument $page */
+        $page = $this->documentManager->create('page');
+        $page->setTitle('Example page');
+        $page->setLocale('en');
+        $page->setStructureType('test_page');
+        $page->setResourceSegment('/example-page-123');
+        $page->setParent($this->documentManager->find($this->sessionManager->getContentPath('sulu_io')));
+
+        $page->getStructure()->bind([
+            'title' => 'Example page',
+            'template' => 'test_page',
+            'url' => '/test',
+            'animals' => [$snippet->getUuid()],
+        ]);
+
+        $this->documentManager->persist($page, 'en');
+        $this->documentManager->publish($page, 'en');
+        $this->documentManager->flush();
+
+        $this->pageReferenceProvider->updateReferences($page, 'en', 'test');
+        $this->getEntityManager()->flush();
+
+        /** @var Reference[] $references */
+        $references = $this->referenceRepository->findBy(['referenceContext' => 'test']);
+        $this->assertCount(1, $references);
+        self::assertSame($snippet->getUuid(), $references[0]->getResourceId());
+    }
+
+    public function testUpdateUnpublishedReferences(): void
+    {
+        /** @var SnippetDocument $snippet */
+        $snippet = $this->documentManager->create('snippet');
+        $snippet->setStructureType('animal');
+        $snippet->setTitle('Example snippet');
+        $snippet->setWorkflowStage(WorkflowStage::PUBLISHED);
+        $this->documentManager->persist($snippet, 'en');
+
+        /** @var PageDocument $page */
+        $page = $this->documentManager->create('page');
+        $page->setTitle('Example page');
+        $page->setLocale('en');
+        $page->setStructureType('test_page');
+        $page->setResourceSegment('/example-page-123');
+        $page->setParent($this->documentManager->find($this->sessionManager->getContentPath('sulu_io')));
+
+        $page->getStructure()->bind([
+            'title' => 'Example page',
+            'template' => 'test_page',
+            'url' => '/test',
+            'animals' => [$snippet->getUuid()],
+        ]);
+
+        $this->documentManager->persist($page, 'en');
+        $this->documentManager->publish($page, 'en');
+        $this->documentManager->flush();
+
+        $this->documentManager->unpublish($page, 'en');
+        $this->documentManager->flush();
+        $this->documentManager->clear();
+
+        /** @var PageDocument $page */
+        $page = $this->documentManager->find($page->getUuid(), 'en');
+        $page->setStructureType(null);
```

Comment:
@alexander-schranz we have to set it here to `null` forcefully, otherwise I'm not sure how I can manage to get the `page` from the `website` context in a test 🤔

Reply:
@Prokyonn theoretical you can boot the website kernel after you shutdown the admin kernel:

```php
        static::ensureKernelShutdown();
        $websiteKernel = static::createKernel(['sulu.context' => SuluKernel::CONTEXT_WEBSITE);
```

Reply:
Yep that worked, thank you 👍